### DISCUSSION TITLE
Generate multiple JSON schema files that refer to each others

### DISCUSSION THREAD

**Discussion by robotechdy** [pydantic/pydantic#4168] [Question]
Firstly, thanks for creating such an awesome library, `pydantic`.

I am using it to generate JSON schemas. Then, I can convert JSON schemas to class definitions in other languages. Finally, I can build a protocol generator to replace the Protocol Buffer (which I personally dislike).

Thanks to `pydantic`, it is pretty easy to generate JSON schemas with nested `pydantic` class definitions. 
Suppose I define `class Foo` in a file `foo.py`, and define `class Bar` in another file `bar.py`, `class Bar` has a member `foo: Foo`. Then, `Bar.schema_json()` can generate one JSON schema for `Bar` with a nested JSON schema for `Foo` in it.

Now, my question is how would I generate two JSON schemas (e.g. exported into two `.json` files), where:
1. One JSON Schema file only contains the JSON schema for `Foo`.
2. Another JSON Schema file only contains the JSON schema for `Bar`.
3. JSON schema in the file of `Bar` contains a reference pointing to JSON schema in the file of `Foo`.

If the scenario above works, we can extend it to generate a folder of JSON schema files that has one-to-one mappings to the original python files defining the original `pydantic` classes.

**Comment by robotechdy** [comment]
Still no comment for this? I thought this is very interesting and useful

**Comment by samuelcolvin** [comment]
Sorry for the slow reply.

The short answer is that I don't think there's much appetite to support this directly from pydantic.

The best approach would be:
* edit `schema()` for each model (you could use a common base class) to include the file name in the schema under some item name that JSON Schema and your tool ignore
* build a custom post-processor script to go through the generate schema and save it in different files

Hope that helps.

**Reply by robotechdy** [reply]
@samuelcolvin 
Thanks for your suggestion! 
I am not sure if I understand your idea completely. 
Here belows are some code snippets I tried to implement your idea.

apple.py
```
from pydantic import BaseModel, Field

class Apple(BaseModel):
    size: int
    weight: int
    file_path: str = Field(default=__file__, const=True)
```

root.py
```
from pydantic import BaseModel, Field
from path.to.apple import Apple

class Root(BaseModel):
    location: str
    apple: Apple
    file_path: str = Field(default=__file__, const=True)

if __name__ == '__main__':
    print(Root.schema_json())
```

Then, if I do `python root.py`, I got a JSON schema
```
    {"title": "Root", "type": "object",
     "properties": {"location": {"title": "Location", "type": "string"},
                    "apple": {"$ref": "#/definitions/Apple"},
                    "file_path": {"title": "File Path",
                                  "const": "/path/to/root.py",
                                  "type": "string"}}, "required": ["location", "apple"], "definitions": {
        "Apple": {"title": "Apple", "type": "object", "properties":
            {"size": {"title": "Size", "type": "integer"},
             "weight": {"title": "Weight", "type": "integer"},
             "file_path": {"title": "File Path",
                           "const": "/path/to/apple.py",
                           "type": "string"}},
                  "required": ["size", "weight"]}}}
```
If I create a `post_processor.py`, I can definitely split this JSON schema into two files: `/path/to/root.json` and `/path/to/apple.json` (and let `root.json` has a reference pointing to `apple.json`). Did the code snippets above implement your idea correctly? I have a few questions about your idea.

question-1. I did not implement a base class like `base_model.py` below
```
class BaseModelWithPath(BaseModel):
        file_path: str = Field(default=__file__, const=True)
```
Since this results `file_path=='/path/to/base_model.py'` all the time.
So what is your original idea about the "common base class" you mentioned?

question-2. You suggested I read and modify the schema generating code of `schema()` (which affects `schema_json()`) function?

**Comment by samuelcolvin** [comment]
You have the basic idea, I mean you can define a new `MyBaseMode` which inherits from `BaseModel`, and defined a custom `schema` method. Then use that in `Apple`, `Root` etc.

Beyond that, I can't really help, I'm on my phone right now.

**Reply by robotechdy** [reply]
@samuelcolvin 
Sorry for the late reply. I am busy doing my current job recently.
So I think this still requires me to hack the source code to make some custom `schema` method here. I am still very interested in this. I will hack related source code to make this work later, when I am free from my current job.

**Reply by robotechdy** [reply]
@samuelcolvin 
Hi, I am back to check (or build) this feature. 
I know pydantic v2 was released. Does v2 include any updates related (or helpful) to this?

**Comment by FunkMonkey** [comment]
I had the same requirement and worked myself around this using a custom `GenerateJsonSchema`:

```python
# this is a factory function creating a new class, so that we can pass information into it, since pydantic is itself instantiating
# the passed GenerateJsonSchema class
def create_generate_json_schema_with_externals(defs_to_remove: List[str], refs_to_rename: Dict[str, str]) -> type[GenerateJsonSchema]:
    """
    Creates a custom json schema generator, which excludes specific definitions and allows the altering of refs
    to them.
    """

    class GenerateJsonSchemaWithExternals(GenerateJsonSchema):

        def generate_inner(self, schema: Any) -> JsonSchemaValue:
            # Skipping the creation of the definitions we want to remove.
            # Since generate_inner is called recursively for the whole tree, this will also
            # skip the processing of all child properties and thus other definitions.
            if "model_name" in schema and schema["model_name"] in defs_to_remove:
                return {}

            return super(GenerateJsonSchemaWithExternals, self).generate_inner(schema)

        def generate(self, schema: CoreSchema, mode: JsonSchemaMode = "validation") -> JsonSchemaValue:
            result = super(GenerateJsonSchemaWithExternals, self).generate(schema, mode)

            # Since generate_inner returns {}, it will still create stub definitions with a title.
            # We'll just remove them here.
            if "$defs" in result:
                defs = result["$defs"]
                for def_to_remove in defs_to_remove:
                    if def_to_remove in defs:
                        del defs[def_to_remove]

            return result

        def _build_definitions_remapping(self) -> _DefinitionsRemapping:

            remapping = super(GenerateJsonSchemaWithExternals, self)._build_definitions_remapping()
            json_remapping = remapping.json_remapping

            # Remap the refs, so they can point to external files.
            for key, ref in json_remapping.items():
                if ref in refs_to_rename:
                    json_remapping[key] = JsonRef(refs_to_rename[ref])

            return remapping

    return GenerateJsonSchemaWithExternals
```

Usage: 
```python
# first schema, that we want to link to
schema = SomeExternalDef.model_json_schema() # save this to some json

# second schema, that uses the first
defs_to_remove = ["SomeExternalDef"]
refs_to_rename = { "#/$defs/SomeExternalDef": "./path-to-external-schema.json#/$defs/SomeExternalDef" }
GenerateJsonSchemaWithExternals = create_generate_json_schema_with_externals(defs_to_remove, refs_to_rename)
schema = SomeModel.model_json_schema(schema_generator=GenerateJsonSchemaWithExternals)
```

I hope this helps anyone!

**Reply by koliyo** [reply]
Thanks! The `defs_to_remove` is very useful!

However, I think the `refs_to_remove` functionality is not needed if using the `ref_template` parameter as described here, https://docs.pydantic.dev/latest/concepts/json_schema/#customizing-the-refs-in-json-schema
